File: backoffice/FondaDomain/Domain/Invoice.cs
using System;
using System.Runtime.Serialization;

namespace com.ds201625.fonda.Domain
{
    /// <summary>
    /// Representa la factura de una cuenta
    /// </summary>
    public class Invoice : BaseEntity
    {
        #region Fields

        /// <summary>
        /// Pago al que la factura pertenece
        /// </summary>
        private Payment _payment;

        ///// <summary>
        ///// Cuenta a la que la factura pertenece
        ///// </summary>
        //private Account _account;

        /// <summary>
        /// Profile a la que la factura pertenece
        /// </summary>
        private Profile _profile;

        /// <summary>
        /// Fecha de pago de la cuenta
        /// </summary>
        private DateTime _date;

        /// <summary>
        /// Monto toal a pagar de la cuenta
        /// </summary>
        private float _total;

        /// <summary>
        /// IVA o tax de la cuenta
        /// </summary>
        private float _tax;

        /// <summary>
        /// Estado de la cuenta
        /// </summary>
        private InvoiceStatus _status;
        
        /// <summary>
        /// Moneda 
        /// </summary>
        private Currency _currency;

        /// <summary>
		/// El numero unico de la factura
		/// </summary>
        private int _number;
        #endregion

        #region Constructors
        /// <summary>
        /// Constructor
        /// </summary>
        public Invoice() : base() { }

        /// <summary>
        /// Constructor de Factura
        /// </summary>
        /// <param name="payment">Pago de la factura</param>
        /// <param name="profile">Perfil de la factura</param>
        /// <param name="total">Total de la factura</param>
        /// <param name="tax">Impuesto de la factura</param>
        /// <param name="currency">Tipo de Moneda</param>
        public Invoice(Payment payment, Profile profile,
         float total, float tax, Currency currency, int number, InvoiceStatus status) 
            : base()
        {
            

            this._payment = payment;
            this._profile = profile;
            this._date = DateTime.Now;
            this._total = total;
            this._tax = tax;

            this._status = status;

            this._currency = currency;
            this._number = number;
        }

        /// <summary>
        /// Constructor de Factura
        /// </summary>
        ///<param name="id">Id de la factura</param>
        /// <param name="payment">Pago de la factura</param>
        /// <param name="profile">Perfil de la factura</param>
        /// <param name="total">Total de la factura</param>
        /// <param name="tax">Impuesto de la factura</param>
        public Invoice( int id,Payment payment, Profile profile, 
            float total, float tax, int number)
            : base()
        {
            this.Id = id;
            this._payment = payment;
            this._profile = profile;
            this._date = DateTime.Now;
            this._total = total;
            this._tax = tax;
            this._status = new GeneratedInvoiceStatus();
            this._number = number;
        }

        /// <summary>
        /// Constructor de Factura sin propina
        /// </summary>
        /// <param name="payment">Pago de la factura</param>
        /// <param name="profile">Perfil de la factura</param>
        /// <param name="total">Total de la factura</param>
        /// <param name="tax">Impuesto de la factura</param>
        /// <param name="currency">Tipo de Moneda</param>
        public Invoice(Payment payment, Profile profile,
            float total, float tax, Currency currency, int number)
            : base()
        {
            this._payment = payment;
            this._profile = profile;
            this._date = DateTime.Now;
            this._total = total;
            this._tax = tax;
            this._status = new GeneratedInvoiceStatus();
            this._currency = currency;
            this._number = number;

        }

        #endregion

        #region Properties
        /// <summary>
        /// Obtiene o asigna una moneda a la factura
        /// </summary>
        [DataMember]
        public virtual Currency Currency
        {
            get { return _currency; }
        }


        ///// <summary>
        ///// Obtiene o asigna una cuenta a la factura
        ///// </summary>
        //[DataMember]
        //public virtual Account Account
        //{
        //    get { return _account; }
        //}

        /// <summary>
        /// Obtiene o asigna un restaurant a la cuenta
        /// </summary>
        //[DataMember]
        //public virtual Restaurant Restaurant
        //{
        //    get { return _restaurant; }
        //}

        /// <summary>
        /// Obtiene o asigna un perfil a la cuenta
        /// </summary>
        [DataMember]
        public virtual Profile Profile
        {
            get { return _profile; }
        }

        /// <summary>
        /// Obtiene o asigna la fecha de pago de la cuenta
        /// </summary>
        [DataMember]
        public virtual DateTime Date
        {
            get { return _date; }
        }

        /// <summary>
        /// Obtiene o asigna el monto total a pagar de la cuenta
        /// </summary>
        [DataMember]
        public virtual float Total
        {
            get { return _total; }
        }

        /// <summary>
        /// Obtiene o asigna el IVA
        /// </summary>
        [DataMember]
        public virtual float Tax
        {
            get { return _tax; }
        }

        /// <summary>
        /// Obtiene o asigna el numero de la factura
        /// </summary>
        public virtual int Number
        {
            get { return _number; }
        }

        /// <summary>
        /// Obtiene o asigna el status de la cuenta
        /// </summary>
        public virtual InvoiceStatus Status
        {
            get { return _status; }
            set { _status = value; }
        }
        
        /// <summary>
        /// Obtiene o asigna un pago
        /// </summary>
        public virtual Payment Payment
        {
            get { return _payment; }
        }
        #endregion

        /// <summary>
        /// Cambia el estado actual de la factura.
        /// </summary>
        public virtual void changeStatus()
        {
            _status = _status.Change();
        }
    }
}

File: frontoffice/app/src/main/java/com/ds201625/fonda/domains/entities/Invoice.java
package com.ds201625.fonda.domains.entities;

import com.ds201625.fonda.domains.BaseEntity;

import java.util.Date;

/**
 * Created by Adri on 06/05/2016.
 */

/*
 Clase quee representa la factura del restaurant
 */
public class Invoice extends BaseEntity {

    /**
     * Float que define la propina de la cuenta
     */
    private float tip;

    /**
     * Date que define la fecha de pago de la cuenta
     */
    private Date date;

    /**
     *Float que define el impuesto del pago de la cuenta
     */
    private float tax;

    /**
     * Float que define el total de la cuenta
     */
    private float total;

    /**
     * Restaurant que define el restaurant asociado a la cuenta
     */
    private Restaurant restaurant;

    /**
     *Profile que define el perfil asociado a la cuenta
     */
    private Profile profile;

    /**
     *Currency que deifne el tipo de moneda a usar en la cuenta
     */
    private Currency currency;

    /**
     *Payment que define el pago de la cuenta
     */
    private Payment payment;

    /**
     *Account que define la cuenta
     */
    private Account account;


    /**
   Contructor de la clase factura
     */
    public Invoice() {
        super();
    }

    public Invoice(float tip, float tax, float total, Restaurant restaurant, Date date,
        Profile profile) {
        this.tip = tip;
        this.tax = tax;
        this.total = total;
        this.restaurant = restaurant;
        this.date = date;
        this.profile= profile;
    }

    public Invoice(float tip, float tax, float total, Restaurant restaurant, Date date,
        Profile profile, Currency currency,Payment payment, Account account) {
        this.tip = tip;
        this.tax = tax;
        this.total = total;
        this.restaurant = restaurant;
        this.date = date;
        this.profile= profile;
        this.currency= currency;
        this.payment= payment;
        this.account= account;
    }

    /**
     * Metodo que obtiene la propina de la cuenta
     * @return propina
     */
    public float getTip() {
        return tip;
    }

    /**
     * Metodo que asigna la propina de la cuenta
     * @param tip
     */
    public void setTip(Float tip) {
        this.tip = tip;
    }

    /**
     * Metodo que obtiene el impuesto de la cuenta
     * @return impuesto
     */
    public float getTax() {return tax;}

    /**
     * Metodo que asigna el impuesto de la cuenta
     * @param tax
     */
    public void setTax(Float tax) {
        this.tax = tax;
    }

    /**
     * Metodo que obtiene el total de la cuenta
     * @return total
     */
    public float getTotal() {
        return total;
    }

    /**
     * Metodo que asigna el total de la cuenta
     * @param total
     */
    public void setTotal(Float total) {
        this.total = total;
    }

    /**
     * Metodo que obtiene la fecha del pago de la cuenta
     * @return fecha
     */
    public Date getDate() {
        return date;
    }

    /**
     * Metodo que asigna la fecha del pago de la cuenta
     * @param date
     */
    public void setDate(Date date) {
        this.date = date;
    }

    /**
     * Metodo que obtiene el Restaurant asociado a la cuenta
     * @return pestaurante
     */
    public Restaurant getRestaurant() {
        return restaurant;
    }

    /**
     * Metodo que asigna un restaurant a la cuenta
     * @param restaurant
     */
    public void setRestaurant(Restaurant restaurant) {this.restaurant = restaurant;}

    /**
     * Metodo que obtiene un perfil asociado a la cuenta
     * @return perfil
     */
    public Profile getProfile() {
        return profile;
    }

    /**
     * Metodo que asigna un perfil a la cuenta
     * @param profile
     */
    public void setProfile(Profile profile) {this.profile = profile;}

    /**
     * Metodo que obtiene la moneda del pago de la cuenta
     * @return moneda
     */
    public Currency getCurrency() {
        return currency;
    }

    /**
     * Metodo que asigna una moneda a la cuenta
     * @param currency
     */
    public void setCurrency(Currency currency) {
        this.currency = currency;
    }

    /**
     * Metodo que obtiene el pago de la cuenta
     * @return pago
     */
    public Payment getPayment() {
        return payment;
    }

    /**
     * Metodo que asigna el pago de la cuenta
     * @param payment
     */
    public void setPayment(Payment payment) {
        this.payment = payment;
    }

    /**
     * Metodo que obtiene la Cuenta
     * @return cuenta
     */
    public Account getAccount() {
        return account;
    }

    /**
     * Metodo que asigna la cuenta
     * @param account
     */
    public void setAccount(Account account) {
        this.account = account;
    }
}

